src/runtime/proxy-component.ts

```diff
@@ -62,6 +62,21 @@ export const proxyComponent = (Cstr: d.ComponentConstructor, cmpMeta: d.Componen
       prototype.attributeChangedCallback = function (attrName: string, _oldValue: string, newValue: string) {
         plt.jmp(() => {
           const propName = attrNameToPropName.get(attrName);
+
+          // the attr changed callback runs prior to Stencil's connectedCallback, which may also attempt to
+          //  unshadow lazy properties. If we attempt to unshadow in both places
+          // to cover the case where an attr was set inline on the non-upgraded element and then the property
+          // was programmatically set which will be handled here or the case where the attr was not set in
+          // which case the connectedCallback will catch and unshadow.
+          // https://developers.google.com/web/fundamentals/web-components/best-practices#lazy-properties
+          // we should think about whether or not we actually want to be reflecting the attributes to properties
```

Comment:
Can we add the TODO that I originally proposed here? I'd like to start tracking things like this in our planning tools (I may have created a placeholder ticket 😉).
```suggestion
          // TODO(STENCIL-16) we should think about whether or not we actually want to be reflecting the attributes to properties
```

Or did you want to have that conversation as a part of this ticket?